Add _forward_jacobian private function for faster Jacobians

ricdigi opened the pull request:
#### References to other Issues or PRs
This PR is based on PR #25801 by @brocksam. To understand the reasons behind this new function
I suggest to directly take a look at the original PR discussion.

The current goal of this imlpmentation is just to offer a faster option to the classic jacobian
when handling very big system of equations.

While the initial aim of the function was to be public, in the latest modifications introduced 
with this PR the function becomes private. Moreover it doesn't rely anymore (for now) on 
the CseExpr class (introduced in PR #24649). 

#### Brief description of what is fixed or changed

- Refactored the `forward_jacobian` function to be private.
- Enhanced `forward_jacobian` to handle expressions containing `dynamicsymbols`.
- Optimized the transformation process from DAG to canonical expression tree.
- Updated tests to include expressions with `dynamicsymbols`.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* simplify
  * Refactored `forward_jacobian` function to be private and handle `dynamicsymbols`.
  * Optimized DAG to canonical expression tree transformations.
<!-- END RELEASE NOTES -->

brocksam committed:
Add new CseExpr class, update cse function to return CseExpr

brocksam committed:
Add tests for new CseExpr class

brocksam committed:
Add CseExpr to sympy and sympy.simplify namespaces

brocksam committed:
Add CseExpr class to docs

brocksam committed:
Update docstring to remove reference to past behaviour

brocksam committed:
Remove as_tuple method as simple to implement with tuple cast

brocksam committed:
Rename CseExpr attributes to reduced_exprs and replacements

brocksam committed:
Update expected return type to use tuples not lists

brocksam committed:
Rename constructor method to from_reduced_exprs

brocksam committed:
Merge branch 'master' into cse-expr

brocksam committed:
Revert backward incompatible change to return type

brocksam committed:
Handle edge case that cse_separate returns list not tuple

brocksam committed:
Add note about exceptional handling of cse_separate

brocksam committed:
Add test case with CSEd expression as expr argument

brocksam committed:
Add support for passing CseExpr as expr argument

brocksam committed:
Add CseExpr to docstring, as CseExpr doctest

brocksam committed:
Implement forward_jacobian function for faster Jacobians using CSE

brocksam committed:
Add basic tests for forward_jacobian

ricdigi committed:
Refactor forward_jacobian() function

This commit implements the following modifications:
- Make the forward_jacobian function private.
- Enhance forward_jacobian to handle expressions containing dynamicsymbols.
- Optimize the transformation process from DAG to canonical expression tree.
- Remove the option to output as CseExpr object.
- Update tests to include expressions with dynamicsymbols.

ricdigi committed:
Merge master into feature-branch and resolve conflicts

- Merged latest changes from master into feature-branch
- Resolved conflicts in sympy/utilities/tests/test_lambdify.py
- Updated feature-branch with new features from master

ricdigi committed:
Removed extra newline in test_cse_diff.py

ricdigi commented:
@tjstienstra @moorepants please let me know what you think.

oscarbenjamin commented on sympy/simplify/cse_main.py:
```diff
@@ -40,6 +40,195 @@
+            The reduced expression(s).
+        replacements_mapping : ``Dict[Symbol, Expr]``
+            Pairs of replacements present in ``exprs``.
+        as_list : ``bool``, default is ``True``
+            Whether the argument passed to ``reduced_exprs`` should be converted
+            to a ``list``. With the default of ``True``, if an iterable is
+            passed to ``reduced_exprs`` then it is converted to a ``list``, or
+            if a single ``Expr`` is passed then it is wrapped into a single-item
+            ``list``. If ``False``, then the argument passed to
+            ``reduced_exprs`` is left as is, meaning that the value returned by
+            the ``reduced_exprs`` property is not necessarily a ``list``. This
+            parameter is similar to the ``list`` parameter of the ``cse``
+            function.
```
This parameter is messy. It should be for the caller to pass a list or not or there should be a `CseExprList` type or something to represent the list case.

The simplest solution is to say that the caller should always pass a list.

oscarbenjamin commented:
Please also respond to the review comments from the previous PR:
https://github.com/sympy/sympy/pull/25801#discussion_r1362479274

oscarbenjamin commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,190 @@
+            replacement_symbol, partial_derivative = add_to_cache(subexpr.diff(free_symbol))
+            absolute_derivative += partial_derivative * absolute_derivative_mapping.get(free_symbol, zeros)
+
+        # Modification to manage dynamicsymbols
+        if free_symbols == {dynamicsymbols._t}:
+            absolute_derivative_mapping[symbol] = ImmutableDenseMatrix([[subexpr.diff(sub) for sub in wrt]])
+            continue
+        absolute_derivative_mapping[symbol] = ImmutableDenseMatrix([[add_to_cache(a)[0] for a in absolute_derivative]])
```
Why is this special case needed?

It would be better not to have a special case for dynamicsymbols.

moorepants replied:
Yes, this should simply work with functions. The concept of "dynamic symbols" should remain in the physics vector package.

ricdigi replied:
@moorepants Unfortunately this is the only way found to make it work with expressions containing dynamicsymbols for now.  But indeed if the function is placed in cse_diff it should not contain this special case.

moorepants replied:
A "dynamic symbol" is simply a `Function('x')(Symbol('y'))`. If this function stays in this location, then we need to make it general and handle any SymPy function, which is a super set of `Function('x')(dynamicsymbols._t)`.

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,190 @@
+
+def _forward_jacobian(
+    expr: ImmutableDenseMatrix,
+    wrt: ImmutableDenseMatrix,
```
This seems a bit restrictive. It would be better to accept anything that quacks like a matrix, i.e. has rows and columns with entries that are sympy expressions and `wrt` should be any iterable of sympy variables (symbols or functions).

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,190 @@
+    Expressions often contain repeated subexpressions. If and expression is
+    represented as a tree structure then multiple copies of these subexpressions
+    will be present in the expanded form of the expression. During
+    differentiation these repeated subexpressions will be repeatedly and
```
```suggestion
    differentiation these repeated subexpressions will be repeatedly
```

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,190 @@
+    represented as a tree structure then multiple copies of these subexpressions
+    will be present in the expanded form of the expression. During
+    differentiation these repeated subexpressions will be repeatedly and
+    differentiated multiple times, resulting in repeated and wasted work.
```
```suggestion
    differentiated multiple times, resulting in duplicate computations.
```

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,190 @@
+
+    symbols = expr.free_symbols
+    replacement_symbols = numbered_symbols(
+        prefix='_z',
```
Worth making this more unique to avoid clashes if the user can't set it.

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,190 @@
+        The vector to be differentiated.
+    wrt : ``ImmutableDenseMatrix``
+        The vector with respect to which to do the differentiation.
+    as_cse_expr : ``bool``
```
I don't see where this is used in the code.

ricdigi committed:
Revert commits relative to CseExpr class

Commits relative to the CseExpr class have been removed.
However, to maintain trace of past contribution, the revert
has been performed by manually deleting lines of code.

ricdigi committed:
Second Revert of commits relative to CseExpr class

Commits relative to the CseExpr class have been removed.
However, to maintain trace of past contribution, the revert
has been performed by manually deleting lines of code.

ricdigi commented:
### Update
I discovered a bug in the current implementation:

When derivatives of common subexpressions with respect to one of the variables in the `wrt` list are present in the input expression, the Jacobian is not computed correctly. This occurs because, during the CSE step, the subexpression (which is the argument of the derivative) is replaced with a new symbol. Consequently, when computing the derivative of this term with respect to any variable in the `wrt` vector, it is incorrectly evaluated to zero instead of being calculated accurately.

Here is an example to illustrate this issue:

```python
from sympy import ImmutableDenseMatrix, Symbol, Function, Derivative, simplify
from sympy.simplify.cse_diff import _forward_jacobian

# Define the symbols
x = Symbol('x')
y = Symbol('y')
z = Symbol('z')

# Define the custom functions
k = Function('k')(x, y)
f = Function('f')(k, z)

# Define the expression and the variables with respect to which the Jacobian is computed
expr = ImmutableDenseMatrix([Derivative(f + k, x) + k])
wrt = ImmutableDenseMatrix([x])

# Compute the Jacobian using the custom implementation
jacobian_ric = _forward_jacobian(expr, wrt)

# Compute the Jacobian using the classical method
jacobian_cla = expr.jacobian(wrt)

# Calculate the difference between the two Jacobians to identify discrepancies
diff = simplify(jacobian_ric - jacobian_cla)

print(diff)

# Output:
# Matrix([[-Derivative(f(k(x, y), z), k(x, y))*Derivative(k(x, y), (x, 2)) - Derivative(f(k(x, y), z), (k(x, y), 2))*Derivative(k(x, y), x)**2 - Derivative(k(x, y), x) - Derivative(k(x, y), (x, 2))]])
```

A possible solution suggested by @tjstienstra, would be to post process the CSEd expression by substituting back the original expressions when that expression is the argument of a derivative. 

Another solution could be using a post processing as shown in this [comment](https://github.com/sympy/sympy/pull/24649#issuecomment-1766578770), and redesigning the function to work with that. 

@moorepants what do you think?

ricdigi committed:
Refactor _forward_jacobian() for flexibility and robustness

- Handle Derivative terms and dynamic symbols in expressions.
- Add regression test for Derivative term handling.
- Refactor test_lambdify for improved code quality.

ricdigi commented:
I decided to refactor the forward jacobian function with another implementation I developed. This implementation should work with dynamicsymbols, and correct the problem of Derivative term handling explained in the previous post:

> A possible solution suggested by @tjstienstra, would be to post process the CSEd expression by substituting back the original expressions when that expression is the argument of a derivative.

@moorepants @tjstienstra please let me know your thought about it. Moreover, could you suggest other examples to test the robustness of this implementation? 

What do you think instead of the input format checks, are they appropriate?

P.S. I ran the performance benchmark on this new implementation, and it has approximately the same performances. I will do a more quantitative and detailed report soon.

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,188 @@
+from collections import Counter
+
+
+def postprocess(repl, reduced):
```
```suggestion
def _postprocess(repl, reduced):
```

Let's make all these private.

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,188 @@
+
+def _forward_jacobian(expr, wrt):
+    """
+        Returns the Jacobian matrix produced using a forward accumulation
```
indented too far

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,188 @@
+    sub_rep = dict(replacements)
+    for i, ik in enumerate(precomputed_fs):
+        sub_dict = {j: sub_rep[j] for j in ik}
+        sub_rep[rep_sym[i]] = sub_rep[rep_sym[i]].xreplace(sub_dict)
```
It would be nice to have a function that returns the cse'd Jacobian and then a second function that is a light wrapper that subs back in the originals to give the full Jacobian.

moorepants commented on sympy/simplify/cse_diff.py:
```diff
@@ -0,0 +1,188 @@
+        sub_rep[rep_sym[i]] = sub_rep[rep_sym[i]].xreplace(sub_dict)
+
+    J = {key: expr.xreplace(sub_rep) for key, expr in J.items()}
+    J = SparseMatrix(None, J)
```
Seems a bit odd to return a `SparseMatrix`. Shouldn't it return a normal SymPy Matrix?

moorepants replied:
Should be the same type as the input.

oscarbenjamin replied:
A normal SymPy matrix is internally sparse and already has sparse matrix multiplication. The same is true fo SparseMatrix which has the same implementation internally as Matrix. There is never any reason to use SparseMatrix now but it is kept for compatibility.

I suspect that the reason for not using it here is because there is this which is a bit slower than normal multiplication:
https://github.com/sympy/sympy/blob/e30c119507fa64d4d8f7e78c09bf1649abb2e9b3/sympy/polys/matrices/sdm.py#L1453-L1471
https://github.com/sympy/sympy/blob/e30c119507fa64d4d8f7e78c09bf1649abb2e9b3/sympy/polys/matrices/sdm.py#L1497-L1503
This is what is used for normal matrix multiplication even though `sdm_matmul` would otherwise be faster. The faster routine assumes that `0*expr` evaluates to `0` for all expressions which is not true for all expressions in SymPy. The faster routine is used for domains like `QQ[x,y]` but not when the domain is EXRAW as is the default for matrices that are not purely rational numbers.

Rather than using other types here or reimplementing what `DomainMatrix` already does it would be better just to add a method to Matrix to compute a matrix product without trying to handle things like `0*oo`. The implementation is already there but it is not used by default to preserve existing behaviour like:
```python
In [1]: Matrix([[x, oo]]) * Matrix([0, 0])
Out[1]: [nan]
```
Also though if the expressions in the matrix are at all large than I doubt that all the zero multiplies are going to make a lot of difference to the total time for the matrix product. Possibly multiplication by zero can be optimised as well.

ricdigi replied:
> The implementation is already there but it is not used by default to preserve existing behaviour like

Could you point me where this method is implemented? 

@oscarbenjamin could you also explain me more practically what objects and methods to use instead of what I am using now? Especially relative to the matrix multiplication. Thank you in advance.

oscarbenjamin replied:
> could you also explain me more practically what objects and methods to use instead of what I am using now?

Can you first demonstrate that there is a good reason for not just using Matrix here?

I realise that what you are doing involves sparse matrices but it also likely involves very large complicated expressions which tends to mean that the benefit of the sparse approach is lost in the weeds (a few extra zero multiplies makes no noticeable difference in context).

oscarbenjamin replied:
> Could you point me where this method is implemented?

I linked it above already.

moorepants commented on sympy/simplify/cse_main.py:
```diff
@@ -764,11 +764,11 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
+         earlier in this list might show up in subexpressions later in this
+         list.
+     reduced_exprs : list of SymPy expressions
+         The reduced expressions with all of the replacements above.
```
Seems like the indentation was already correct, no?

ricdigi committed:
Improved input postprocessing

- Made the auxiliary functions private
- Using regex to check which replacement symbols are present
in an expression instead of subtracting original symbols set.

ricdigi committed:
Split core logic and add wrapper functions

- Modified Docstring to comply to standards
- Updated Test for _cse_diff

ricdigi commented:
## Update

### Split into core and wrappers

In the latest commit sI have implemented the following modifications.

I split the original function into three parts. One contains the main functionality and is private(`_forward_jacobian_core`). It takes a CSEd expression as input and outputs the CSEd Jacobian, thus a replacements list and a reduced Matrix.

There a two wrappers, one (`_forward_jacobian_norm_in_dag_out`) which takes a normal expression and wrt and outputs outputs the CSEd Jacobian, thus a replacements list and a reduced Matrix.

And another (`forward_jacobian`) which takes a normal expression and wrt and outputs the Jacobian Matrix with the original symbols.

@moorepants @tjstienstra what do you think of this organization? 

### Use of Sparse Matrices

Regarding the use of SparseMatrices, I had started using them because I saw a performance improvement, however, the performance improvement might have been casued by the removal of this behavior instead:

```markdown
In [1]: Matrix([[x, oo]]) * Matrix([0, 0])
Out[1]: [nan]
```

And not from the use of sparse matrices. 

Thus it might be indeed useful to try use normal matrices  and use a product that  doesn’t try to handle things like `0*oo` and see if the performances remain the same.

Please let me know what do you think about this.

### Performances

I tried swapping the current forward_jacobian impementation inside physics/mechancis/linearization and it seem to pass all the pytest tests.

Moreover I timed the linearization time for the bike model in physics/mechancis/tests/test_kane3.py, using the same benchamrk procedure (cache clearing, etc.) and these are the results:

```markdown
forward_jacobian, Average Time: 41.306592109600004, Sub Times: {'total': [41.309234499, 41.325472485000006, 41.69748056, 41.13626438699998, 41.064508617]}
jacobian_classic, Average Time: 129.6768561508, Sub Times: {'total': [125.689425846, 128.525281506, 125.81091524799996, 136.09811826700002, 132.26053988700005]}
```

Thus the function seem to have a **3x speed up** wrt the classic jacobian.

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    Postprocess the CSE output to remove any CSE replacement symbols from the arguments
+    of Derivative terms.
+    """
```
Best to make complete docstrings here now, so we know what the input/output exactly should be.

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,244 @@
+        The matrix of expressions with respect to which to differentiate the reduced
+        expression.
+
+    """
```
Add a "Returns" section to these docstrings so you fully specify the outputs.

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,244 @@
+        J[(i, j)] += value
+
+    J = SparseMatrix(l_red, l_wrt, J)
+    J = reduced_expr[0].__class__(J)
```
Add unit tests that show different input matrix types are properly output.

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,244 @@
+    return replacements, J, precomputed_fs
+
+
+def _forward_jacobian_norm_in_dag_out(expr, wrt):
```
add docstring

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,244 @@
+    expr : Matrix
+        The vector to be differentiated.
+
+    wrt : Matrix, list, or tuple
```
The type should be an iterable of something. What is the something? Only Symbol and Function()(Symbol)?

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,244 @@
+        sub_dict = {j: sub_rep[j] for j in ik}
+        sub_rep[rep_sym[i]] = sub_rep[rep_sym[i]].xreplace(sub_dict)
+
+    J = {key: expr.xreplace(sub_rep) for key, expr in J.todok().items()}
```
```suggestion
    jacobian = {key: expr.xreplace(sub_rep) for key, expr in J.todok().items()}
```

moorepants replied:
Best to make variables pep8 unless necessary for mathematical clarity. Here J isn't needed for math.

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,244 @@
+    J = SparseMatrix(l_red, l_wrt, J)
+    J = reduced_expr[0].__class__(J)
+
+    return replacements, J, precomputed_fs
```
Add unit tests to make sure these outputs are structured same as cse() (those outputs are a bit complex I think and it should try to follow same pattern)

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,244 @@
+    return replacements, J
+
+
+def forward_jacobian(expr, wrt):
```
should be private

ricdigi commented:
# Update
I compared the execution times of my function when using either `sdm_matmul_exraw()` or `sdm_matmul` for non-EXRAW expressions, and confirmed that checking for possible NaN results in the multiplication was causing the slow performance.

Ensuring the correctness of the result is of utmost importance, so I must perform the check for possible NaN results, instead of just skipping zero values..

**OBSERVATION:** In my function, every time sparse matrix multiplication is performed, the second matrix is always the same (Matrix C). However, at each iteration, a new row is added to C. This means some elements of C are redundantly checked when using `sdm_matmul_exraw()`. Therefore, I implemented a matrix multiplication function with a check for NaN which take advantage of this, avoiding redundant checks. The function accepts two extra sets that store information about which elements of the two matrices lead to NaN when multiplied by zero. These sets can be updated and reused across multiple iterations.

The sets are computed using a helper function `_check_nan()`.

@oscarbenjamin

I am not 100% sure if `sdm_matmul_exraw()` performs other safety checks that I might be neglecting. Please let me know your thoughts on this.

I think these helper functions could still be optimized. I implemented them quickly to see if, with this change, I could maintain good performance while performing the multiplication with the appropriate safety checks.

Please let me know if you think this implementation is viable, or if I am neglecting something.

**Helper functions:**

```python
from collections import Counter
from sympy import S, nan

def _check_nan(A, nan_idx, n):
    """
    Check if any element of Matrix A multiplied by zero doesn't return zero.
    For each element which doesn't return zero, update the nan_idx set
    to store either the row index (n=0) or column index (n=1).
    """
    
    zero = S.Zero

    if n == 0:
        for (i, k), value in A.items():
            if zero * value != zero:
                nan_idx.add(i)
    elif n == 1:
        for (k, j), value in A.items():
            if zero * value != zero:
                nan_idx.add(j)

    return nan_idx

def _dok_matmul_with_nan_handling(A, B, nan_idx_A, nan_idx_B, rows_A, cols_B):
    """
    Sparse matrix multiplication handling possible undefined results (e.g., 0 * infinity).
    It uses the indexes of elements in A and B that produce NaN values, stored in nan_idx_A and nan_idx_B.
    """
    
    C = Counter()
    C_int = Counter()

    # Handle NaN propagation
    nan_rows = set(nan_idx_A)
    nan_cols = set(nan_idx_B)

    for i in nan_rows:
        for j in range(cols_B):
            C[(i, j)] = nan
    for j in nan_cols:
        for i in range(rows_A):
            C[(i, j)] = nan

    # Standard matrix multiplication to form C_int
    for (i, k), A_value in A.items():
        for (k2, j), B_value in B.items():
            if k == k2:
                C_int[(i, j)] += A_value * B_value

    # Combine C and C_int to form the final result C
    for key, value in C_int.items():
        C[key] = value

    return C
```

**Comparison with `matmul()` from `polys/matrices/sdm.py`**

```python
import sympy as sm
from sympy import SparseMatrix
from sympy.polys.matrices.sdm import SDM, sdm_matmul_exraw
from sympy import EXRAW
import time

# Define symbols
inf = sm.oo
x = sm.symbols('x')
y = sm.symbols('y')

# Define the first matrix in DOK format
A = {(2,0): inf, (33, 45): x, (33, 15): x}  # Only element (2, 0) is non-zero and equal to infinity
A_s = SDM.from_dok(A, (1000, 1000), EXRAW)

# Define the second matrix in DOK format
B = {(0,1): inf, (0,86): inf, (45, 33): y, (15, 33): 2*y**2} 
B_s = SDM.from_dok(B, (1000, 1000), EXRAW)

print("Matrix A:", A)
print("Matrix B:", B)

# TIME matmul
start_time = time.time()
C = A_s.matmul(B_s)
end_time = time.time()
print(end_time - start_time)

C = SparseMatrix(None, C.to_dok())

# TIME new helper function
rows_A = 1000  # Number of rows in A
cols_B = 1000  # Number of columns in B
nan_idx_A = set()
nan_idx_B = set()

start_time = time.time()
nan_idx_A = _check_nan(A, nan_idx_A, 0)
nan_idx_B = _check_nan(B, nan_idx_B, 1)
C1 = _dok_matmul_with_nan_handling(A, B, nan_idx_A, nan_idx_B, rows_A, cols_B)
end_time = time.time()
print(end_time - start_time)

C1 = SparseMatrix(None, C1)

# TEST CORRECTNESS
print(C == C1)
```

### Results:

@tjstienstra, @moorepants

I have implemented these new helper functions and benchmarked their performance using the usual procedure. Here are some results:

**Average execution time of test_kane3.py out of 5 runs**: 47.14 sec for forward jacobian **vs** 129.67 sec for the classic Jacobian.

**Execution time on the bicycle model (1 run)**: 13.66 sec for forward jacobian **vs** 155.05 sec for the classic Jacobian.

**Average execution times for pendulum with increasing number of links:**

| n_links | exec_time (s) |
| --- | --- |
| 1 | 0.042 |
| 2 | 0.17 |
| 3 | 0.32 |
| 4 | 0.54 |
| 5 | 0.85 |
| 6 | 1.52 |
| 7 | 2.04 |
| 8 | 2.78 |
| 9 | 5.28 |
| 10 | 6.83 |
| 11 | 9.73 |

ricdigi commented:
### Updated to do list

**To Do - General**

- [ ]  Make forward_jacobian private
- [ ]  Create new PR to let users try linearization performances
- [ ]  Assert correctness of matrix multiplication (through comm. feedback)

**To Do - Dcumentation** 

- [ ]  Add docstrings for the helper functions
- [ ]  Complete docstrings for forward_jacobian
- [ ]  clearly specify paramters and returns

**To Do - Testing**

- [ ]  Add unit test for each helper function
- [ ]  Add unit tests for different types of input
- [ ]  Test the output is in correct cse() format

**To Do - Style**

- [ ]  Ensure all variable are pep8

ricdigi committed:
Updated helper function for matrix multiplication

- The helper function now check for possible NaN results
instead of completely skipping multiplicaiton by zero

- Made forward_jacobian private

- Added tests from classic jacobian tests in test_cse_diff.py

oscarbenjamin commented:
Is the overhead of `sdm_matmul_exraw` actually measurable in the context of a real calculation? I would expect that with expressions of any significant size it would not be noticeable.

You keep showing numbers but not comparisons of things that would be meaningful like A vs B. I have no context for understanding if your table of timings is good or bad. The overhead of `sdm_matmul_exraW` might be 0.00001% of time in a real calculation but I have no idea from the numbers that you show. It is not even clear if you have measured its time in the context of an operation where other nontrivial operations are involved.

You seem to be doing good work but the way that you explain what you have done and what the results are leaves me not having any idea what the significance of any of it is.

oscarbenjamin commented:
> Therefore, I implemented a matrix multiplication function with a check for NaN which take advantage of this, avoiding redundant checks. The function accepts two extra sets that store information about which elements of the two matrices lead to NaN when multiplied by zero. These sets can be updated and reused across multiple iterations.

I think that a better approach here is just to raise an error if the matrix contains anything that is an infinity or a nan (if it is necessary to handle this) and then use a matrix multiplication routine that does not need to check for these things.

Rather than making a separate function here for matrix multiplication just change the Matrix code so that we can use the existing routines directly. I've shown a diff for this below. It would be worth putting that in a separate PR if planning to use it.

First remove this check:
https://github.com/sympy/sympy/blob/a9a6f150383de85a70a19e91e88bca40ceb093ba/sympy/polys/matrices/sdm.py#L1470-L1471
There is only one place in the codebase that calls this function so move the check there instead.

Then add some method to `Matrix` to call through to `sdm_matmul()` but not using the `_exraw` function.  There is already a `Matrix.multiply()` method so it could have an argument like `skipnan=True` or something.

There is no point in using `SparseMatrix` as I've said before. The internal representation of `SparseMatrix` and `Matrix` is exactly the same. They both have an internal DomainMatrix:
```python
In [3]: M = Matrix([[1, 2], [3, 4]])

In [4]: S = SparseMatrix([[1, 2], [3, 4]])

In [5]: M._rep
Out[5]: DomainMatrix({0: {0: 1, 1: 2}, 1: {0: 3, 1: 4}}, (2, 2), ZZ)

In [6]: S._rep
Out[6]: DomainMatrix({0: {0: 1, 1: 2}, 1: {0: 3, 1: 4}}, (2, 2), ZZ)
```
If you want to construct a Matrix efficiently from a dict then you can use `Matrix.from_dok`.

Here is a suggested diff to make a fast matrix multiplication available for for EXRAW that skips all the nan checking. It also changes `sdm_matmul` to use `K.sum` (aka `Add(*terms)`) for faster addition of expression trees like `sdm_matmul_exraw` does. I also changed it to use `defaultdict` which I think speeds it up a little bit e.g. for `randMatrix(100)*randMatrix(100)`. You can access the faster routine like:
```python
In [1]: M1, M2 = Matrix([[oo]]), Matrix([[0]])

In [2]: M1*M2
Out[2]: [nan]

In [3]: M1.multiply(M2)
Out[3]: [nan]

In [4]: M1.multiply(M2, fast=True)
Out[4]: [0]
```
```diff
diff --git a/sympy/matrices/matrixbase.py b/sympy/matrices/matrixbase.py
index 3dcdfc6c83..ff4aa771c3 100644
--- a/sympy/matrices/matrixbase.py
+++ b/sympy/matrices/matrixbase.py
@@ -2815,7 +2815,7 @@ def __mul__(self, other):
 
         return self.multiply(other)
 
-    def multiply(self, other, dotprodsimp=None):
+    def multiply(self, other, dotprodsimp=None, fast=False):
         """Same as __mul__() but with optional simplification.
 
         Parameters
@@ -2842,10 +2842,12 @@ def multiply(self, other, dotprodsimp=None):
             if self.shape[1] != other.shape[0]:
                 raise ShapeError(f"Matrix size mismatch: {self.shape} * {other.shape}.")
 
-            m = self._eval_matrix_mul(other)
-
-            if isimpbool:
-                m = m._new(m.rows, m.cols, [_dotprodsimp(e) for e in m])
+            if fast:
+                m = self._eval_matrix_mul_sparse(other)
+            else:
+                m = self._eval_matrix_mul(other)
+                if isimpbool:
+                    m = m._new(m.rows, m.cols, [_dotprodsimp(e) for e in m])
 
             return m
 
diff --git a/sympy/matrices/repmatrix.py b/sympy/matrices/repmatrix.py
index e20781eefc..c2cde6807b 100644
--- a/sympy/matrices/repmatrix.py
+++ b/sympy/matrices/repmatrix.py
@@ -351,6 +351,10 @@ def _eval_add(self, other):
     def _eval_matrix_mul(self, other):
         return classof(self, other)._fromrep(self._rep * other._rep)
 
+    def _eval_matrix_mul_sparse(self, other):
+        A, B = self._rep.unify(other._rep)
+        return classof(self, other)._fromrep(A.matmul_sparse(B))
+
     def _eval_matrix_mul_elementwise(self, other):
         selfrep, otherrep = self._rep.unify(other._rep)
         newrep = selfrep.mul_elementwise(otherrep)
diff --git a/sympy/polys/matrices/_dfm.py b/sympy/polys/matrices/_dfm.py
index c2f6f16922..0032f548f4 100644
--- a/sympy/polys/matrices/_dfm.py
+++ b/sympy/polys/matrices/_dfm.py
@@ -425,6 +425,9 @@ def matmul(self, other):
         shape = (self.rows, other.cols)
         return self._new(self.rep * other.rep, shape, self.domain)
 
+    def matmul_sparse(a, b):
+        return a.to_sdm().matmul_sparse(b.to_sdm()).to_ddm()
+
     # XXX: For the most part DomainMatrix does not expect DDM, SDM, or DFM to
     # have arithmetic operators defined. The only exception is negation.
     # Perhaps that should be removed.
diff --git a/sympy/polys/matrices/ddm.py b/sympy/polys/matrices/ddm.py
index 77d02b8ad5..49b52e3fcb 100644
--- a/sympy/polys/matrices/ddm.py
+++ b/sympy/polys/matrices/ddm.py
@@ -703,6 +703,9 @@ def matmul(a, b):
         ddm_imatmul(c, a, b)
         return c
 
+    def matmul_sparse(a, b):
+        return a.to_sdm().matmul_sparse(b.to_sdm()).to_ddm()
+
     def mul_elementwise(a, b):
         assert a.shape == b.shape
         assert a.domain == b.domain
diff --git a/sympy/polys/matrices/domainmatrix.py b/sympy/polys/matrices/domainmatrix.py
index b91bef314d..2e785b16ed 100644
--- a/sympy/polys/matrices/domainmatrix.py
+++ b/sympy/polys/matrices/domainmatrix.py
@@ -1600,6 +1600,10 @@ def matmul(A, B):
         A._check('*', B, A.shape[1], B.shape[0])
         return A.from_rep(A.rep.matmul(B.rep))
 
+    def matmul_sparse(A, B):
+        A._check('*', B, A.shape[1], B.shape[0])
+        return A.from_rep(A.rep.matmul_sparse(B.rep))
+
     def _scalarmul(A, lamda, reverse):
         if lamda == A.domain.zero:
             return DomainMatrix.zeros(A.shape, A.domain)
diff --git a/sympy/polys/matrices/sdm.py b/sympy/polys/matrices/sdm.py
index 0e0685f5d6..a59dc473d0 100644
--- a/sympy/polys/matrices/sdm.py
+++ b/sympy/polys/matrices/sdm.py
@@ -778,6 +778,21 @@ def __rmul__(a, b):
         else:
             return NotImplemented
 
+    def _matmul(A, B, sparse=False):
+        if A.domain != B.domain:
+            raise DMDomainError
+        m, n = A.shape
+        n2, o = B.shape
+        if n != n2:
+            raise DMShapeError
+
+        if not sparse and A.domain.is_EXRAW:
+            C = sdm_matmul_exraw(A, B, A.domain, m, o)
+        else:
+            C = sdm_matmul(A, B, A.domain)
+
+        return A.new(C, (m, o), A.domain)
+
     def matmul(A, B):
         """
         Performs matrix multiplication of two SDM matrices
@@ -811,14 +826,10 @@ def matmul(A, B):
         {0: {0: 8}, 1: {0: 2, 1: 3}}
 
         """
-        if A.domain != B.domain:
-            raise DMDomainError
-        m, n = A.shape
-        n2, o = B.shape
-        if n != n2:
-            raise DMShapeError
-        C = sdm_matmul(A, B, A.domain, m, o)
-        return A.new(C, (m, o), A.domain)
+        return A._matmul(B, sparse=False)
+
+    def matmul_sparse(A, B):
+        return A._matmul(B, sparse=True)
 
     def mul(A, b):
         """
@@ -1450,7 +1461,7 @@ def sdm_matvecmul(A, B, K):
     return C
 
 
-def sdm_matmul(A, B, K, m, o):
+def sdm_matmul(A, B, K):
     #
     # Should be fast if A and B are very sparse.
     # Consider e.g. A = B = eye(1000).
@@ -1467,28 +1478,19 @@ def sdm_matmul(A, B, K, m, o):
     # Aik and Bk are both nonzero. In Python the intersection of two sets
     # of int can be computed very efficiently.
     #
-    if K.is_EXRAW:
-        return sdm_matmul_exraw(A, B, K, m, o)
-
     C = {}
     B_knz = set(B)
     for i, Ai in A.items():
-        Ci = {}
-        Ai_knz = set(Ai)
-        for k in Ai_knz & B_knz:
+        Ci_terms = defaultdict(list)
+        for k in Ai.keys() & B_knz:
             Aik = Ai[k]
             for j, Bkj in B[k].items():
-                Cij = Ci.get(j, None)
-                if Cij is not None:
-                    Cij = Cij + Aik * Bkj
-                    if Cij:
-                        Ci[j] = Cij
-                    else:
-                        Ci.pop(j)
-                else:
-                    Cij = Aik * Bkj
-                    if Cij:
-                        Ci[j] = Cij
+                Ci_terms[j].append(Aik * Bkj)
+        Ci = {}
+        for j, Cij_terms in Ci_terms.items():
+            Cij = K.sum(Cij_terms)
+            if Cij:
+                Ci[j] = Cij
         if Ci:
             C[i] = Ci
     return C

ricdigi commented:
@oscarbenjamin thanks for the nice explanation, I will probably proceed how you are suggesting. To be more complete on my explanation, and my findings:

I compared the execution times of different `forward_jacobian()` implemntations using different matrix multiplication implementations: `sdm_matmul` for the non-EXRAW case, `sum_matmul_exraw`, and my custom matrix multiplication. These implementations can be found in this [repository](https://github.com/ricdigi/jacobian_benchmark) in the implementations folder, namely named `forward_jacobian_sdm.py`, `forward_jacobian_sdm_non_exraw`, and `forward_jacobian_final`.

The input symbolic expressions used for benchmarking are based on real use cases, such as the bicycle model.

Here are the results:

**Benchmark 1**: Input expressions derived from pendulum with **n** number of links
<img width="835" alt="Screenshot 2024-08-01 at 11 48 52" src="https://github.com/user-attachments/assets/d6b8fc68-565e-4b4b-ac27-3343a55e85cb">


**Benchmark 2** Input expressions from the bicycle model (J.P Meijaard et al. -  doi: 10.1098/rspa.2007.1857)
Here's the summary in a table:

| Implementation                | Input Size | WRT Size | Total Time (s)  |
|-------------------------------|------------|----------|-----------------|
| jacobian_classic              | 3          | 38       | 130.68          |
| forward_jacobian_sdm          | 3          | 38       | 28.42           |
| forward_jacobian_sdm_non_exraw| 3          | 38       | 11.49           |
| forward_jacobian_final        | 3          | 38       | 14.29           |

oscarbenjamin commented:
> The input symbolic expressions used for benchmarking are based on real use cases, such as the bicycle model.

Okay, so it does make a significant difference. It would be worth profiling to see what it is that takes the time. It might be that something else can be improved.

It is not clear to me how to run a particular example problem with a particular method from your repo. Also the `forward_jacobian_sdm_non_exraw` method does not seem to be there (maybe you need to push).

oscarbenjamin commented:
Also I notice that you are calling the operation repeatedly in a loop for the timings. I can see that you are clearing the cache but this is not sufficient. There are many different levels of caching in sympy and not all of them can be cleared. The correct way to time an operation as discussed previously is that the operation is run once in a subprocess. You can launch many subprocesses in a loop to collect multiple timings and average them but you should not repeat the same operation in a loop within the same process.

It would be useful to add a helper function to SymPy for timing an operation in a subprocess and collecting timings. Something like a `timeit_sympy()` function.

oscarbenjamin commented:
> I will probably proceed how you are suggesting.

Let's add the `mul_sparse` function to Matrix in a separate PR and get that merged. If it makes a noticeable difference here then it will be needed in other situations as well.

I timed the `forward_jaobian_sdm` method with `sizes=(12,)` (although it says `input_size=13`) for some reason and I got this:
```console
$ python main.py 
[{'implementation': 'forward_jacobian_sdm', 'input_size': 13, 'wrt_size': 51, 'total_time': 39.5295601910002}]
forward_jacobian_sdm - Input Size: 13, Total Time: 39.5295601910002, Sub Times: {'total': [39.5295601910002]}
```
So it takes about 40 seconds which is a little slower than shown in your plot but seems consistent (presumably this computer is slower).

Now I run this under a profiler to see what takes the time:
```console
$ python -m cProfile -s cumulative main.py > t.time
```
The profiler output is:
```
$ head -n 30 t.time 
[{'implementation': 'forward_jacobian_sdm', 'input_size': 13, 'wrt_size': 51, 'total_time': 95.7039181580003}]
forward_jacobian_sdm - Input Size: 13, Total Time: 95.7039181580003, Sub Times: {'total': [95.7039181580003]}
         696872483 function calls (637317443 primitive calls) in 628.204 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    765/1    0.014    0.000  628.215  628.215 {built-in method builtins.exec}
        1    0.002    0.002  628.215  628.215 main.py:1(<module>)
        1    0.001    0.001  626.631  626.631 benchmark.py:49(run_benchmark_pendulum)
        6    1.598    0.266  557.774   92.962 forward_jacobian_sdm.py:5(forward_jacobian_sdm)
16306943/4113259   27.668    0.000  490.506    0.000 cache.py:69(wrapper)
        1    0.030    0.030  462.104  462.104 utils.py:21(warm_up_function)
    10823    6.295    0.001  401.395    0.037 sdm.py:1497(sdm_matmul_exraw)
10141039/3806375   20.450    0.000  380.626    0.000 numbers.py:1899(__mul__)
6752180/3376266    5.567    0.000  370.415    0.000 decorators.py:58(__sympifyit_wrapper)
3372628/3372567    5.370    0.000  366.565    0.000 numbers.py:1459(__mul__)
2305441/2293679   35.051    0.000  365.701    0.000 operations.py:52(__new__)
3451705/3450505    3.573    0.000  361.004    0.000 decorators.py:224(_func)
3463060/3417893    3.478    0.000  358.470    0.000 decorators.py:99(binary_op_wrapper)
3372532/3372471    4.339    0.000  355.454    0.000 numbers.py:465(__mul__)
  3416419    4.608    0.000  349.003    0.000 expr.py:216(__mul__)
2226906/2218908   68.294    0.000  287.691    0.000 mul.py:195(flatten)
725439/207195    2.121    0.000  120.911    0.001 expr.py:3573(diff)
730247/211589    6.263    0.000  120.458    0.001 function.py:1900(_derivative_dispatch)
735602/211473   13.847    0.000  117.414    0.001 function.py:1232(__new__)
        1    0.004    0.004   95.704   95.704 benchmark.py:19(time_function)
  4451640   35.768    0.000   83.068    0.000 mul.py:465(_gather)
75859/31224    0.099    0.000   70.164    0.002 function.py:1891(_dispatch_eval_derivative_n_times)
        1    0.000    0.000   68.812   68.812 models.py:18(generate_input_pendulum)
```
Note the everything is slower under the profiler so it gives much bigger times for everything. Usually though the relative times shown for different operations are still meaningful.

The profiler says that most of the time is spend in `sdm_matmul_exraw` but it also says that it is being called 10823 times. I can't believe that 10823 matrix products are needed in this operation.

How does computing the Jacobian end up needing this many matrix products?

It seems to be computing one matrix product for each step in the CSE representation but I can't see why that would be necessary at all. I don't understand it but the algorithm here is clearly wildly different from what I was imagining.

The way that this is done in protosym (which is much faster for this example) does not involve any matrix products. The CSE representation is just used for computing derivatives of the subexpressions in the matrix.

I think this is why I have been confused about your focus on sparse matrix multiplication. I was imagining that you were computing the Jacobian and then using it for a matrix product at the end in which case I would be very surprised if the `exraw` business would have any effect on the overall computing time. It seems that actually you are doing something very different...

tjstienstra commented:
> How does computing the Jacobian end up needing this many matrix products?
> 
> It seems to be computing one matrix product for each step in the CSE representation but I can't see why that would be necessary at all. I don't understand it but the algorithm here is clearly wildly different from what I was imagining.

As I understand, the algorithm uses matrices to store intermediate results. I think that the overall computation is similar to what was done by Sam. In [this document}(https://www.overleaf.com/read/kbhysvrrvtpb#1b4934) I've started to write out the equations (the notation may not be perfect, but it should give a rough idea). It is not finished yet, but I currently don't have more time to complete it

oscarbenjamin commented:
> I think that the overall computation is similar to what was done by Sam

Maybe it is. It looks different from a typical implementation of forward differentiation. I guess it is because it is computing derivatives for all variables simultaneously. At some point I tried reverse mode accumulation since it is supposedly more efficient for computing multiple derivatives simultaneously but I found that in the symbolic case it did not seem to have the same benefits as in the automatic differentiation case. I found that I got simpler expressions in the same amount of time just by using forward accumulation once for each differentiation variable. I didn't explore it in any depth though.

@ricdigi the "warmup function" and running the timings in a loop makes the timings shown not necessarily representative due to caching. Somewhere I mentioned the idea of a warmup function but it needs to be a function that is not computing the exact result that you are going to time subsequently. The idea would be to do something like:
```python
# First warmup with a small input
e = myfunc(1)

# Now measure once with the actual input:
start = time.time()
myfunc(1000) # big input which is slow
end  = time.time()
print(end - start)
```
The point of the warmup is to allow any one-time initialisation such as delayed imports etc to occur so that they are not in the timing results. However this needs to be done without running the exact same operation that are going to time. What you are doing is this:
```python
clear_cache() # Why is this happening here?

# First warmup by calling the function 5 times
for _ in range(5): # a loop is not needed here
    e = myfunc(1000) # The argument here should be different

# It would make more sense to clear the cache at this point before timings...

# Now time making the exact same function call repeatedly:
start = time.time()
for _ in range(ntimes): # This should not use a loop
    myfunc(1000)  # this exact call is already warm cache because of warmup above
end  time.time()
print(end - start)
```
The cache makes a big difference to subsequent runs e.g.:
```python
In [5]: M = randMatrix(100, 1)*x + randMatrix(100, 1)*y

In [6]: %time ok = M.jacobian([x, y])
CPU times: user 112 ms, sys: 3.78 ms, total: 115 ms
Wall time: 115 ms

In [7]: %time ok = M.jacobian([x, y])
CPU times: user 56.8 ms, sys: 20 μs, total: 56.9 ms
Wall time: 56 ms
```
Only the first call to Jacobian there is the one whose time we should be interested in. The way you are timing it means that the first call happens in `warmup_function()` and is not measured.

ricdigi committed:
Updated _process_cse function and relative unit test

- Combined three helper functions in one (_process_cse)
- Improved _process_cse docstring
- Added unit test for _process_cse in test_cse_diff.py

ricdigi committed:
Refactored to use standard matrix multiplication methods

- Standard matrix mult. routines are now used
- Completed docstrings
- Added unit test for input and output formats of forward_jacobian
- Matrix variable name refactored for pep8

ricdigi commented:
## Update

Following your suggestions, I proceeded to refactor the function to use standard matrix multiplication routines. 

At the current state, the function works correctly and already offers fast performance for some of the tested application cases:

**Benchmark 1**: Input expressions from the bicycle model (J.P Meijaard et al. -  doi: 10.1098/rspa.2007.1857)

| Implementation     | Input Size | WRT Size | Total Time (s) |
|--------------------|------------|----------|----------------|
| jacobian_classic   | 3          | 38       | 130.68         |
| _forward_jacobian  | 3          | 38       | 26.34          |

**Benchmark 2**: Execution time of test_kane3.py: 42.81 sec for forward jacobian vs 129.67 sec for the classic jacobian.

NOTE: The timings were still produced through my benchmark routine as in the previous posts. Despite giving qualitative insight on the performance, the benchmark should be improved as suggested by @oscarbenjamin.

The performance could be further improved by introducing a new method in the matrix class to skip NaN checks, as discussed in previous posts. I plan to implement this in a separate PR as suggested.

Given the performance improvements this function can already bring, I believe this PR is ready to be merged.
@moorepants @tjstienstra

moorepants commented:
@oscarbenjamin in our weekly GSoC meeting I suggested that Riccardo focus on getting this to a mergeable state regardless of performance, focusing on correctness. He can then work on speeding up things and looking into the matrix multiplication modifications in a subsequent PR. We are also merging this as a completely private module so we can iterate more freely before making it public or making it available via `Matrix.jacobian()`.

tjstienstra commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,263 @@
+    if not (reduced_expr[0].shape[0] == 1 or reduced_expr[0].shape[1] == 1):
+        raise TypeError("``expr`` must be a row or a column matrix")
+
+    if not isinstance(wrt, (MatrixBase, list, tuple)):
```
You can use `iterable` from `sympy.utilities.iterables`

tjstienstra commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,263 @@
+        sub_dict = {j: sub_rep[j] for j in ik}
+        sub_rep[rep_sym[i]] = sub_rep[rep_sym[i]].xreplace(sub_dict)
+
+    jacobian = {key: expr.xreplace(sub_rep) for key, expr in jacobian[0].todok().items()}
```
Why not use xreplace directly on the `jacobian[0]` matrix?

tjstienstra commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,263 @@
+                                        for j, s in enumerate(rep_sym) if s in fs and (diff_value := r.diff(s)) != 0})
+
+    precomputed_fs = [
+        {symbol for symbol in s.free_symbols if re.compile(r'x\d+').fullmatch(symbol.name)}
```
Why use a regular expression match? This may result in errors when someone uses a different generator in cse. I would expect that you can create a set with all replacement symbols and take the intersection here.

tjstienstra commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,263 @@
+def _forward_jacobian(expr, wrt):
+    r"""
+    Function to compute the Jacobian of an input Matrix of expressions through forward accumulation.
+    Takes a sympy Matrix of expression (expr) as input and an iterable of variables (with_respect_to) with respect to
```
```suggestion
    Takes a sympy Matrix of expression (expr) as input and an iterable of variables (wrt) with respect to
```

ricdigi committed:
Improved _cse_diff.py from review comments

- Used `iterable` from `sympy.utilities.iterables` for cleaner type checks.
- Applied `xreplace` directly on `jacobian[0]`
- Replaced regex filtering with set intersection for safer symbol handling.
- Corrected error in Docstring

ricdigi committed:
Corrected indentation error in cse_main.py

ricdigi committed:
Merge branch 'master' into forward-jacobian-update

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+
+
+def _process_cse(replacements, reduced_expressions):
+    r"""
```
```suggestion
    """
```

raw string only necessary if it contains some invalid characters.

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+def _process_cse(replacements, reduced_expressions):
+    r"""
+    This function is designed to postprocess the output of a common subexpression elimination (CSE)
+    operation. Specifically, it removes any CSE replacement symbols from the arguments of `Derivative` terms in
```
```suggestion
    operation. Specifically, it removes any CSE replacement symbols from the arguments of ``Derivative`` terms in
```

single backticks in RST do nothing, use double backticks for monospaced fonts

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    =======
+
+    processed_replacements : list of (Symbol, expression) pairs
+        Processed replacement list, in the same format of the 'replacements' input list.
```
```suggestion
        Processed replacement list, in the same format of the ``replacements`` input list.
```

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+        Processed replacement list, in the same format of the 'replacements' input list.
+
+    processed_reduced : list of SymPy expressions
+        Processed reduced list, in the same format of the 'reduced_expressions' input list.
```
```suggestion
        Processed reduced list, in the same format of the ``reduced_expressions`` input list.
```

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+from sympy.utilities.iterables import iterable
+
+
+def _process_cse(replacements, reduced_expressions):
```
```suggestion
def _remove_cse_from_derivative(replacements, reduced_expressions):
```

that could be a more informative name. "process" can mean anything

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+
+    repl_dict = dict(replacements)
+    processed_replacements = [(rep_sym, traverse(sub_exp, repl_dict)) for rep_sym, sub_exp in replacements]
+    processed_reduced = [red_exp.__class__([traverse(exp, repl_dict) for exp in red_exp]) for red_exp in reduced_expressions]
```
wrap lines to 79 characters in SymPy

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+
+
+def _forward_jacobian_core(replacements, reduced_expr, wrt):
+    r"""
```
```suggestion
    """
```

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    The function also returns a list of precomputed free symbols for each subexpression,
+    which are useful in the substitution process.
+
+    NOTE: The output Jacobian might not be always exactly in a DAG form, but close to it.
```
You could just not say in "DAG form" and say the output is in common sub expression form.

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    return processed_replacements, processed_reduced
+
+
+def _forward_jacobian_core(replacements, reduced_expr, wrt):
```
```suggestion
def _forward_jacobian_cse(replacements, reduced_expr, wrt):
```

better name? "core" doesn't mean much

moorepants replied:
or `cse_forward_jacobian`

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    ==========
+
+    replacements : list of (Symbol, expression) pairs
+        Replacement Symbols and relative Common Subexpressions that have been replaced during a CSE operation.
```
```suggestion
        Replacement symbols and relative common subexpressions that have been replaced during a CSE operation.
```

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    reduced_expr : list of SymPy expressions
+        The reduced expressions with all the replacements from the replacements list above.
+
+    wrt : Matrix, tuple, or list
```
```suggestion
    wrt : iterable
```

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+        The reduced expressions with all the replacements from the replacements list above.
+
+    wrt : Matrix, tuple, or list
+        Iterable of expressions with respect to which to compute the Jacobian Matrix
```
```suggestion
        Iterable of expressions with respect to which to compute the Jacobian matrix.
```

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    =======
+
+    replacements : list of (Symbol, expression) pairs
+        Replacement Symbols and relative Common Subexpressions that have been replaced during a CSE operation.
```
```suggestion
        Replacement symbols and relative common subexpressions that have been replaced during a CSE operation.
```

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    replacements : list of (Symbol, expression) pairs
+        Replacement Symbols and relative Common Subexpressions that have been replaced during a CSE operation.
+        Compared to the input replacement list, the output one doesn't contain replacement symbols inside
+        Derivative's arguments.
```
```suggestion
        ``Derivative``'s arguments.
```

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+
+    jacobian : list of SymPy expressions
+        The list only contains one element, which is the Jacobian Matrix with elements
+        in reduced form (replacement symbols are present)
```
```suggestion
        in reduced form (replacement symbols are present).
```

moorepants replied:
Why is this a list and not simply the matrix?

moorepants replied:
Also, does this return the same type of matrix as passed in?

ricdigi replied:
> Why is this a list and not simply the matrix?

cse always return the expression in a list

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    The function also returns a list of precomputed free symbols for each subexpression,
+    which are useful in the substitution process.
+
+    NOTE: The output Jacobian might not be always exactly in a DAG form, but close to it.
```
As above, if it isn't guaranteed to be a DAG, we don't need to say this. For example, `cse()` says nothing about a DAG (I think).

moorepants commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,255 @@
+    expr : Matrix
+        The vector to be differentiated.
+
+    with_respect_to : Matrix, list, or tuple
```
```suggestion
    with_respect_to : iterable
```

ricdigi committed:
Refactored _cse_diff with modifications from review

- Corrected Docstrings
- Added new test to check output matrix type
- Corrected code line length

ricdigi commented:
@tjstienstra quick update to say that I have completed the mathematical formulation relative to my forward jacobian implementation, which you started on overleaf. I have included it in my report at: [report](https://github.com/sympy/sympy/wiki/GSoC-2024-Report-Riccardo-Di-Girolamo:-Enhancing-Performance--in-SymPy-Mechanics-and-Vector-Modules)

tjstienstra commented on sympy/simplify/_cse_diff.py:
```diff
@@ -0,0 +1,291 @@
+
+    wrt : iterable
+        Iterable of expressions with respect to which to compute the
+        Jacobian matrix
```
You're slightly inconsistent with the dots at the end of the sentences (some are missing).

ricdigi committed:
Added missing dots in docstrings

ricdigi committed:
Corrected lambdify docstring

- Removed also unnecessary blank line from cse_main.py

tjstienstra reviewed (approved):
The code looks good to me. It seems to give the correct results, is faster than the default jacobian, is readable, and well documented. There are of course several aspects that would be interesting for further research and comparison as pointed out in some of the comments. However, those can addressed in future PRs.

moorepants commented:
We just discussed this in the last GSoC meeting and all issues are addressed. The module and an functions are also private so we can adopt in new PRs. Thanks for all the review.

oscarbenjamin commented:
I'm not sure if the release notes carried over because of the bot error:
https://github.com/sympy/sympy/pull/25801#issuecomment-2312619002

In any case the release notes above are a bit cryptic:



> #### Release Notes
> * simplify
>   
>   * Refactored `forward_jacobian` function to be private and handle `dynamicsymbols`.
>   * Optimized DAG to canonical expression tree transformations.

oscarbenjamin commented:
The release notes can be fixed by just editing them on the wiki.

moorepants commented:
@ricdigi can you manually update the release notes on the wiki?

oscarbenjamin commented:
Can you first say here what the release notes are going to be so that they can be reviewed?

It is unfortunate that release notes are in the OP of the PR and the wiki rather than the codebase because it means that it is easy to review the diff but not notice the release notes when merging. Really though the release notes need to be reviewed just like everything else.

The release notes in the OP here are misdirected because they don't explain anything in a way that would be relevant to users.

moorepants commented:
I'm not sure we need release notes at all for this, as it is a completely private module.

oscarbenjamin commented:
Oh, so is the code added here not actually used anywhere yet?

The PR gh-24649 is showing as having been merged but it looks like all the changes were removed from it. That still seems like a useful PR though.

moorepants commented:
This code is not used yet. The only intention is to use it in https://github.com/sympy/sympy/pull/26952 but even there it will remain private. If there is a desire to make this public we can do so in a new PR.

moorepants commented:
> The PR https://github.com/sympy/sympy/pull/24649 is showing as having been merged but it looks like all the changes were removed from it. That still seems like a useful PR though.

It could be reopened, I suppose, but can always be revived. It closed because @ricdigi built his work on top of that PR with the intention to fix it but ended up reducing the scope. I agree there are ideas and implementations there that we can still integrate and use.

oscarbenjamin commented:
Okay, so there should be no release note for this one then.

Can somone remove all release notes from this or any PRs it was based on?

When the implementation is being used it should still have a release note like "the mechanics module has a new implementation of linearisation and computes equations of motion faster" or something. Probably there should be a user-facing option to control which implementation is being used anyway which would need to be mentioned.

If it is the entire output of a GSOC project there should be something in the release notes in the end but that can be added under "highlights" manually instead.

Also nothing is truly private if it is actually being used because it has the potential to cause bugs and for many the first way to track down bugs from a new release is to go look in the release notes. Major internal changes should still be mentioned there even if the intention is that a user would not need to know about them.

oscarbenjamin commented:
> It could be reopened, I suppose, but can always be revived

Of course and it could have been sitting as a closed PR to be revived. What is confusing now is that it looks like it was merged but all its changes are gone. In the end this PR is so different that it should not have kept the commits from the other one.
